src/repr/src/adt/range.rs

```diff
@@ -169,197 +342,98 @@ impl<'a, const UPPER: bool> RangeBound<'a, UPPER> {
     }
 }
 
-/// Describes the value passed in via `RangeBoundDesc`s.
-#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, PartialOrd)]
-pub enum RangeBoundDescValue<D> {
-    Finite { value: D },
-    Infinite,
-}
-
-impl<'a> From<Datum<'a>> for RangeBoundDescValue<Datum<'a>> {
-    /// Treats `Datum::Null` as an infinite bound (appropriate for PG
-    /// semantics), and all other `Datum`s as finite values.
-    fn from(d: Datum<'a>) -> Self {
-        match d {
-            Datum::Null => RangeBoundDescValue::Infinite,
-            value => RangeBoundDescValue::Finite { value },
+impl<'a, const UPPER: bool> RangeBound<Datum<'a>, UPPER> {
+    /// Create a new `RangeBound` whose value is "infinite" (i.e. None) if `d ==
+    /// Datum::Null`, otherwise finite (i.e. Some).
+    ///
+    /// There is not a corresponding generic implementation of this because
+    /// genericizing how to express infinite bounds is less clear.
+    pub fn new(d: Datum<'a>, inclusive: bool) -> RangeBound<Datum<'a>, UPPER> {
+        RangeBound {
+            inclusive,
+            bound: match d {
+                Datum::Null => None,
+                o => Some(o),
+            },
         }
     }
-}
 
-/// Structures arguments for functions that construct ranges.
-#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, PartialOrd)]
-pub struct RangeBoundDesc<D> {
-    pub inclusive: bool,
-    pub value: RangeBoundDescValue<D>,
-}
+    fn canonicalize(&mut self) {
+        match self.bound {
+            None => {
+                self.inclusive = false;
+            }
+            Some(value) => match value {
+                d @ Datum::Int32(_) => self.canonicalize_inner::<i32>(d),
+                d => unreachable!("{d:?} not yet supported in ranges"),
+            },
+        }
+    }
 
-impl<'a> RangeBoundDesc<Datum<'a>> {
-    /// Create a new `RangeBoundDesc` whose value is infinite if `d ==
-    /// Datum::Null`, otherwise finite.
-    pub fn new(d: Datum<'a>, inclusive: bool) -> RangeBoundDesc<Datum<'a>> {
-        RangeBoundDesc {
-            inclusive,
-            value: d.into(),
+    fn canonicalize_inner<T: RangeOps<'a>>(&mut self, d: Datum<'a>)
+    where
+        <T as TryFrom<Datum<'a>>>::Error: std::fmt::Debug,
+    {
+        if let Some(step) = T::step() {
+            // Upper bounds must be exclusive, lower bounds inclusive
+            if UPPER == self.inclusive {
+                let cur = <T>::unwrap_datum(d);
+                self.bound = cur.checked_add(&step).map(|t| t.into());
```

Comment:
I think a failure of `checked_add` should result in an error instead of converting the upper bound to infinity. For example this is the behavior in PostgreSQL:
```
postgres=# SELECT '[2,2147483647]'::int4range;
ERROR:  integer out of range
LINE 1: SELECT '[2,2147483647]'::int4range;
               ^
postgres=# SELECT '[2,2147483647)'::int4range;
   int4range    
----------------
 [2,2147483647)
(1 row)

```

Reply:
This branch on the other hand has this behavior:
```
materialize=> SELECT '[2,2147483647]'::int4range;
 int4range 
-----------
 [2,)
(1 row)
```

This would probably also be a good test to add to SLT.

Reply:
Nice--ty. Forgot to circle back to that. Will fix and add a test.